docs/AllocatorClaims.md

```diff
@@ -0,0 +1,120 @@
+
+There are two variants of this problem, which may have different solutions.
+In one case, the callee wishes to use an object for the duration of the call.
+In another case, the callee wishes to use an object for a long period.
+The main difference between these is the relative costs.
+For objects held for the duration of a call, the cost of two calls into the allocator to hold and release the object may be prohibitive, whereas this is likely to be negligible for objects held for a long time.
+
+Additional constraints
+----------------------
+
+In addition to solving the basic problem, we have a number of additional constraints on the design space:
+
+ - A solution should be possible to apply at the boundaries when wrapping an existing component in a compartment.
+ - Most objects will not use this mechanism (they will be reachable from a single compartment) and so there should be no overhead when not in use.
+ - We are targeting resource-constrained systems and so the total overhead must be low.
+ - We assume compartments may be malicious and so they should not be able to trick the allocator into creating a data structure that takes a very large amount of time to walk.
+
+Possible approach 0: Explicit copies
+------------------------------------
+
+The first approach is simply to punt on the problem entirely.
+When an OS kernel interacts with userspace memory, it calls explicit helpers to copy data to and from the kernel.
+This is necessary because the userspace process may provide invalid pointers or update the memory map to invalidate pointers while the kernel runs.
+
+In many situations, the same approach could work with CHERIoT.
+We could provide a safe memcpy as a library function that runs with interrupts disabled, checks the source and destination pointers, and then performs the copy (up to some bounded size) and reports whether the copy succeeded.
+This would be sufficient in a lot of cases but is not a very friendly programmer model.
+In particular, it means that any compartment wrapping existing unmodified code would have to either (deep) copy any objects passed in or would have to make invasive changes to the wrapped code.
+
+Note: This approach can be orthogonal to others and so we should implement it anyway.
+
+Possible approach 1: Hazard pointers
+------------------------------------
+
+Hazard pointers provide inspiration for the first possible solution.
+This approach involves having a per-compartment list of held objects that the allocator must consult before freeing an object.
+The allocator would provide an API that allowed a compartment to register a table containing a (bounded) array of held objects and an allocation capability.
+On deallocation, the allocator would need to traverse all such lists to determine whether an object is on any of them.
+If the object is found on a list then the metadata should be updated to mark it as owned by the allocation capability associated with the list.
```

Comment:
Ack; I had a comment here and I'm not sure where it went.  :(

In any case: how is compartment B notified that ownership of the object has transferred from compartment A?  When does either compartment B or the allocator know to free the object?

Reply:
That's part of the problem.  Logically, as soon as you insert something to a hazard list, you logically own it: it lazily counts towards your quota.  You must ensure that you have sufficient quota for it, but you can't easily query its size.